Testnet/Multisig/Multisig/MultisigContract.cs

```diff
@@ -0,0 +1,355 @@
+    {
+        EnsureWalletOnly();
+
+        Assert(IsOwner(address), "Can't remove a non-owner.");
+        Assert(OwnerCount > Required, "Can't have less owners than the minimum confirmation requirement.");
+
+        UnsetOwner(address);
+
+        OwnerCount--;
+    }
+
+    /// <summary>
+    /// Only called within the contract constructor to initialise the owners.
+    /// </summary>
+    /// <param name="addresses">The array of addresses serialised as bytes.</param>
+    private void SetOwners(byte[] addresses)
+    {
+        Address[] addressList = Serializer.ToArray<Address>(addresses);
+
+        foreach (var address in addressList)
+        {
+            SetOwner(address);
+        }
+
+        OwnerCount = (uint)addressList.Length;
+    }
+
+    private void SetOwner(Address address)
+    {
+        State.SetBool($"{OwnerPrefix}:{address}", true);
+
+        Log(new OwnerAddition() { Owner = address });
+    }
+
+    private void UnsetOwner(Address address)
+    {
+        State.SetBool($"{OwnerPrefix}:{address}", false);
+
+        Log(new OwnerRemoval() { Owner = address });
+    }
+
+    /// <summary>
+    /// Submits method and parameter metadata describing a contract call that the multisig contract should invoke once it is sufficiently confirmed.
+    /// </summary>
+    /// <param name="data">An array of method parameters encoded as packed byte arrays. See the <see cref="Transaction"/> struct for further information.</param>
+    /// <returns>The transactionId of the submitted multisig transaction.</returns>
+    /// <remarks>The submitter implicitly provides a confirmation for the submitted transaction.</remarks>
+    public ulong Submit(Address destination, string methodName, byte[] data)
+    {
+        EnsureOwnersOnly();
+
+        TransactionCount++;
+
+        State.SetStruct($"{TransactionPrefix}:{TransactionCount}", new Transaction()
+        {
+            Destination = destination,
+            Executed = false,
+            Value = 0,
+            MethodName = methodName,
+            Parameters = data
+        });
+
+        Log(new Submission() { TransactionId = TransactionCount });
+
+        Confirm(TransactionCount);
+
+        return TransactionCount;
+    }
+
+    /// <summary>
+    /// Called by contract owners to indicate that they accept the proposed method invocation provided via <see cref="Submit"/>.
+    /// </summary>
+    /// <param name="transactionId">The transactionId of the submitted multisig transaction.</param>
+    /// <remarks>Once the confirmation count matches or exceeds <see cref="Required"/> the contract call will be executed.</remarks>
+    public void Confirm(ulong transactionId)
+    {
+        Assert(IsOwner(Message.Sender), "Sender is not authorized to confirm.");
+        Assert(transactionId <= TransactionCount, "Transaction ID does not exist.");
+        Assert(!IsConfirmedBy(transactionId, Message.Sender), "Sender has already confirmed the transaction.");
```

Comment:
These 3 assert calls aren't necessary when this method is called by the `Submit` method. Consider splitting out into a private `ConfirmExecute` method which doesn't have these 3 asserts, as well as a public `Confirm` method, which does these asserts, then calls `ConfirmExecute`.

Also more of a nitpick, might want to use `EnsureOwnersOnly()` in place of `Assert(IsOwner(Message.Sender), ...)`